p2p/src/network/pubsub/pubsub_effectful/p2p_network_pubsub_effectful_effects.rs

```diff
@@ -30,57 +44,89 @@ impl P2pNetworkPubsubEffectfulAction {
                     store.dispatch(P2pNetworkPubsubAction::BroadcastSigned { signature });
                 }
             }
-            P2pNetworkPubsubEffectfulAction::IncomingData {
+            P2pNetworkPubsubEffectfulAction::ValidateIncomingMessages {
                 peer_id,
                 seen_limit,
                 addr,
                 messages,
             } => {
+                let mut valid_messages = Vec::with_capacity(messages.len());
+
                 for message in messages {
-                    let result = if let Some(signature) = message.signature.clone() {
-                        if let Ok(Some(pk)) = originator(&message) {
-                            // the message is mutable only in this function
-                            let mut message = message;
-                            let Ok(data) = encode_without_signature_and_key(&mut message) else {
-                                // should never happen;
-                                // we just decode this message, so it should encode without error
-                                bug_condition!("serialization error");
-                                return;
+                    match validate_message(message, store) {
+                        Ok(valid_msg) => valid_messages.push(valid_msg),
+                        Err(error) => {
+                            let error_msg = match error {
+                                PubSubError::MissingSignature => {
```

Comment:
Could this be done using `thiserror::Error` so that the string generated by the error can be reused and traced back to the enum variant more easily

Reply:
Done, much better now as I don't need to do that match to convert it to a string.